src/OrbitVulkanLayer/TimerQueryPool.h

```diff
@@ -104,62 +111,119 @@ class TimerQueryPool {
+    CHECK(device_to_free_slots_.contains(device));
+    std::vector<uint32_t>& free_slots = device_to_free_slots_.at(device);
+    for (uint32_t slot_index : slot_indices) {
+      CHECK(slot_index < num_timer_query_slots_);
+      const SlotState& current_state = slot_states.at(slot_index);
+      CHECK(current_state == SlotState::kQueryPendingOnGpu);
+      if (!slots_done_reading.contains(slot_index)) {
+        slots_marked_for_reset.insert(slot_index);
+        continue;
+      }
+      slots_done_reading.erase(slot_index);
+      slot_states.at(slot_index) = SlotState::kReadyForQueryIssue;
+      free_slots.push_back(slot_index);
+      VkQueryPool query_pool = device_to_query_pool_.at(device);
+      dispatch_table_->ResetQueryPoolEXT(device)(device, query_pool, slot_index, 1);
+    }
   }
 
- private:
-  enum class SlotState { kReadyForQueryIssue = 0, kQueryPendingOnGpu };
-
-  // Resets an occupied slot to be ready for queries again.
-  // If `rollback_only` is set, it will not call to Vulkan to reset the content of that slot.
-  // This is useful if the slot was retrieved, but the actual query was not yet submitted to
+  // Resets an occupied slot to be ready for queries again. It will *not* call to Vulkan to reset
+  // the content of that slot (in contrast to `ResetQuerySlots`).
+  // This is useful, if the slot was retrieved, but the actual query was not yet submitted to
   // Vulkan (e.g. if on resetting the command buffer).
   //
   // Note that the pool must be initialized using `InitializeTimerQueryPool` before.
   // Further, the given slots must be in the `kReadyForQueryIssue` state, i.e. must be a result
   // of `NextReadyQuerySlot` and must not have been reset yet.
-  void ResetQuerySlotsInternal(VkDevice device, const std::vector<uint32_t>& slot_indices,
-                               bool rollback_only) {
+  void RollbackPendingQuerySlots(VkDevice device, const std::vector<uint32_t>& slot_indices) {
     if (slot_indices.empty()) {
       return;
     }
     absl::WriterMutexLock lock(&mutex_);
     CHECK(device_to_query_slots_.contains(device));
     std::vector<SlotState>& slot_states = device_to_query_slots_.at(device);
+    CHECK(device_to_free_slots_.contains(device));
     std::vector<uint32_t>& free_slots = device_to_free_slots_.at(device);
     for (uint32_t slot_index : slot_indices) {
       CHECK(slot_index < num_timer_query_slots_);
       const SlotState& current_state = slot_states.at(slot_index);
       CHECK(current_state == SlotState::kQueryPendingOnGpu);
       slot_states.at(slot_index) = SlotState::kReadyForQueryIssue;
       free_slots.push_back(slot_index);
-      if (rollback_only) {
-        continue;
-      }
-      VkQueryPool query_pool = device_to_query_pool_.at(device);
-      dispatch_table_->ResetQueryPoolEXT(device)(device, query_pool, slot_index, 1);
     }
   }
 
+ private:
+  enum class SlotState { kReadyForQueryIssue = 0, kQueryPendingOnGpu };
```

Comment:
Couldn't we expand the states here into something like 

{ kReadyForQueryIssue = 0, kQueryPendingOnGpu = 1, kDoneReading = 2 }

and then maintain the current state in the (already existing) device_to_query_slots_ vectors? 

You always have the indices of the slots, so looking up the current state and modifying it is easy and fast (definitely a lot faster than looking up indices in a set).

Reply:
We would need 4 states then, right?
{ kReadyForQueryIssue = 0, kQueryPendingOnGpu = 1, kDoneReading = 2, kResetRequested = 3 }

In general sounds not too bad. I will address this tomorrow.

Reply:
Yes, might be that we need four states. The key idea here would be to "flatten" the state space into one dimension. Right now it's spread out over this enum and two sets, with multiple possible states that are illegal (like having state kReadyForQueryIssue and being in either of the sets, slots being in both sets, etc.). With just the enum the code should be a lot easier to understand and it will be easier to assert correctness.

The one redundancy in terms of state is the "free slots" list, but that exists for performance, and I don't see how we can get rid of that while also being fast.

Reply:
Makes sense!

Reply:
Done.